### PR TITLE
[SQL] fix field home of  #__template_styles table

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #23188.

### Summary of Changes
respect docs & logic 
https://docs.joomla.org/Tables/template_styles

### CODE REVIEW

**Comment by ghost** [comment]
@GarryMoveOut please test.

File: administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE `#__template_styles` DROP INDEX `idx_home`;
+ALTER TABLE `#__template_styles` MODIFY `home` tinyint(1) unsigned DEFAULT 0;
+ALTER TABLE `#__template_styles` ADD UNIQUE `idx_home` ( `home` );
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql]: Don't make it unique per the installation script.

```
CREATE TABLE IF NOT EXISTS `#__template_styles` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `template` varchar(50) NOT NULL DEFAULT '',
  `client_id` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `home` char(7) NOT NULL DEFAULT 0,
  `title` varchar(255) NOT NULL DEFAULT '',
  `params` text NOT NULL,
  PRIMARY KEY (`id`),
  KEY `idx_template` (`template`),
  KEY `idx_home` (`home`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 DEFAULT COLLATE=utf8mb4_unicode_ci AUTO_INCREMENT=9;
```

Comment (alikon) [reply administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql]: fixed, thanks

File: installation/sql/mysql/joomla.sql [hunk 2]
```diff
@@ -1830,7 +1830,7 @@ CREATE TABLE IF NOT EXISTS `#__template_styles` (
   `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
   `template` varchar(50) NOT NULL DEFAULT '',
   `client_id` tinyint(1) unsigned NOT NULL DEFAULT 0,
-  `home` char(7) NOT NULL DEFAULT 0,
+  `home` tinyint(1) NOT NULL DEFAULT 0,
```
Comment (Quy) [inline installation/sql/mysql/joomla.sql]:
```suggestion
  `home` tinyint(1) unsigned NOT NULL DEFAULT 0,
```

File: administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE `#__template_styles` DROP INDEX `idx_home`;
+ALTER TABLE `#__template_styles` MODIFY `home` tinyint(1) unsigned DEFAULT 0;
+ALTER TABLE `#__template_styles` ADD INDEX `idx_home` ( `home` );
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql]:
```suggestion
ALTER TABLE `#__template_styles` ADD INDEX `idx_home` (`home`);
```

File: administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE `#__template_styles` DROP INDEX `idx_home`;
+ALTER TABLE `#__template_styles` MODIFY `home` tinyint(1) unsigned DEFAULT 0;
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/3.9.6-2019-04-15.sql]:
```suggestion
ALTER TABLE `#__template_styles` MODIFY `home` tinyint(1) unsigned NOT NULL DEFAULT 0;
```

**Comment by Quy** [comment]
Maybe the following is an issue when using PatchTester.

Run the Database fix the first time:
```
    Database schema version (3.9.3-2019-02-07) does not match CMS version (3.9.6-2019-04-15).
    Table 'nu7mb_template_styles' should not have index 'idx_home'. (From file 3.9.6-2019-04-15.sql.)
    Table 'nu7mb_template_styles' has the wrong type or attributes for column 'home' with type tinyint(1) unsigned. (From file 3.9.6-2019-04-15.sql.)
```
Then the result, but `idx_home` is not created.
```
Table 'nu7mb_template_styles' does not have index 'idx_home'. (From file 3.9.6-2019-04-15.sql.)
```
Run the Database fix again, then `idx_home` is created, however, this message remains:
```
Table 'nu7mb_template_styles' does not have index 'idx_home'. (From file 3.9.6-2019-04-15.sql.)
```

**Comment by alikon** [comment]
ummm maybe splitting the last "CREATE INDEX"  in a new file .... i'll give it a try

File: administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql [hunk 5]
```diff
@@ -0,0 +1,4 @@
+DROP INDEX "#__template_styles_idx_home";
```
Comment (twister65) [inline administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: This index did not exist in the Joomla PostgreSQL database, I have this error when I apply this patch (I think this is another issue):

> ERROR: relation "gvcms_template_style" does not exist

I suggest this change:
`DROP INDEX IF EXISTS "#__template_styles_idx_home";`

**Review by twister65 — CHANGES_REQUESTED** [review]

File: administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql (hunk 5 above)
Comment (alikon) [reply administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: that index exists https://github.com/joomla/joomla-cms/blob/staging/installation/sql/postgresql/joomla.sql#L1809

Comment (twister65) [reply administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: I can not find `#__template_styles_idx_home` nor `#__template_styles_idx_template` in the update folder:
https://github.com/joomla/joomla-cms/tree/staging/administrator/components/com_admin/sql/updates/postgresql
These indexes are only created in the installation script.

Comment (alikon) [reply administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: even then, they are not removed, so we need to drop old index 
did you test with `IF EXISTS` clause on postgresql ?

Comment (twister65) [reply administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: Yes, my suggestion works well, I tested it successfully.

Comment (alikon) [reply administrator/components/com_admin/sql/updates/postgresql/3.9.6-2019-04-15.sql]: thanks @twister65

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 2572ae5e0f7290b0fad25a18460f87ec0e787734<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24595">issues.joomla.org/tracker/joomla-cms/24595</a>.</sub>

**Comment by ghost** [comment]
@GarryMoveOut please test > https://docs.joomla.org/Testing_Joomla!_patches

**Comment by GarryMoveOut** [comment]
Sorry for not responding for long time. I didn't have time. I will test soon, maybe this in week.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 1116780337f55822c58e3e7245cb6884c6644c9a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24595">issues.joomla.org/tracker/joomla-cms/24595</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24595">issues.joomla.org/tracker/joomla-cms/24595</a>.</sub>

**Comment by HLeithner** [comment]
thx

**Comment by richard67** [comment]
This PR had errors which are fixed now in PR #25299 . Please test.

**Pull request merged**